// objection/lib/utils/knexUtils.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getDialect = getDialect;
exports.isPostgres = isPostgres;
exports.isMySql = isMySql;
exports.isSqlite = isSqlite;
exports.isMsSql = isMsSql;
exports.isKnexQueryBuilder = isKnexQueryBuilder;
exports.isKnexJoinBuilder = isKnexJoinBuilder;
exports.isKnexRaw = isKnexRaw;
exports.isKnexTransaction = isKnexTransaction;

var _builder = require('knex/lib/query/builder');

var _builder2 = _interopRequireDefault(_builder);

var _joinclause = require('knex/lib/query/joinclause');

var _joinclause2 = _interopRequireDefault(_joinclause);

var _raw = require('knex/lib/raw');

var _raw2 = _interopRequireDefault(_raw);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function getDialect(knex) {
  return knex && knex.client && knex.client.dialect || null;
}

function isPostgres(knex) {
  return getDialect(knex) === 'postgresql';
}

function isMySql(knex) {
  return getDialect(knex) === 'mysql';
}

function isSqlite(knex) {
  return getDialect(knex) === 'sqlite3';
}

function isMsSql(knex) {
  return getDialect(knex) === 'mssql';
}

function isKnexQueryBuilder(value) {
  return value instanceof _builder2.default;
}

function isKnexJoinBuilder(value) {
  return value instanceof _joinclause2.default;
}

function isKnexRaw(value) {
  return value instanceof _raw2.default;
}

function isKnexTransaction(knex) {
  return !!getDialect(knex) && typeof knex.commit === 'function' && typeof knex.rollback === 'function';
}
