// objection/lib/transaction.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = transaction;

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _Model = require('./model/Model');

var _Model2 = _interopRequireDefault(_Model);

var _classUtils = require('./utils/classUtils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @returns {Promise}
 */
function transaction() {
  // There must be at least one model class and the callback.
  if (arguments.length < 2) {
    return _bluebird2.default.reject(new Error('objection.transaction: provide at least one Model class to bind to the transaction or a knex instance'));
  }

  var args = new Array(arguments.length);

  for (var i = 0, l = args.length; i < l; ++i) {
    args[i] = arguments[i];
  }

  if (!(0, _classUtils.isSubclassOf)(args[0], _Model2.default) && _lodash2.default.isFunction(args[0].transaction)) {
    var knex = _lodash2.default.first(args);
    args = args.slice(1);

    // If the function is a generator, wrap it using Promise.coroutine.
    if (isGenerator(args[0])) {
      args[0] = _bluebird2.default.coroutine(args[0]);
    }

    return knex.transaction.apply(knex, args);
  } else {
    // The last argument should be the callback and all other Model subclasses.
    var callback = _lodash2.default.last(args);
    var modelClasses = _lodash2.default.take(args, args.length - 1);
    var _i = void 0;

    for (_i = 0; _i < modelClasses.length; ++_i) {
      if (!(0, _classUtils.isSubclassOf)(modelClasses[_i], _Model2.default)) {
        return _bluebird2.default.reject(new Error('objection.transaction: all but the last argument should be Model subclasses'));
      }
    }

    var _knex = _lodash2.default.first(modelClasses).knex();
    for (_i = 0; _i < modelClasses.length; ++_i) {
      if (modelClasses[_i].knex() !== _knex) {
        return _bluebird2.default.reject(new Error('objection.transaction: all Model subclasses must be bound to the same database'));
      }
    }

    // If the function is a generator, wrap it using Promise.coroutine.
    if (isGenerator(callback)) {
      callback = _bluebird2.default.coroutine(callback);
    }

    return _knex.transaction(function (trx) {
      var args = new Array(modelClasses.length + 1);

      for (var _i2 = 0; _i2 < modelClasses.length; ++_i2) {
        args[_i2] = modelClasses[_i2].bindTransaction(trx);
      }

      args[args.length - 1] = trx;

      return _bluebird2.default.try(function () {
        return callback.apply(trx, args);
      });
    });
  }
}

/**
 * @param {Constructor.<Model>|knex} modelClassOrKnex
 * @returns {Promise}
 */
transaction.start = function (modelClassOrKnex) {
  var knex = modelClassOrKnex;

  if ((0, _classUtils.isSubclassOf)(modelClassOrKnex, _Model2.default)) {
    knex = modelClassOrKnex.knex();
  }

  if (!knex || !_lodash2.default.isFunction(knex.transaction)) {
    return _bluebird2.default.reject(new Error('objection.transaction.start: first argument must be a model class or a knex instance'));
  }

  return new _bluebird2.default(function (resolve, reject) {
    knex.transaction(function (trx) {
      resolve(trx);
    }).catch(function (err) {
      reject(err);
    });
  });
};

function isGenerator(fn) {
  return fn && fn.constructor && fn.constructor.name === 'GeneratorFunction';
}
