// objection/lib/queryBuilder/operations/eager/JoinEagerOperation.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _EagerOperation2 = require('./EagerOperation');

var _EagerOperation3 = _interopRequireDefault(_EagerOperation2);

var _RelationJoinBuilder = require('./RelationJoinBuilder');

var _RelationJoinBuilder2 = _interopRequireDefault(_RelationJoinBuilder);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var JoinEagerOperation = function (_EagerOperation) {
  (0, _inherits3.default)(JoinEagerOperation, _EagerOperation);

  function JoinEagerOperation(name, opt) {
    (0, _classCallCheck3.default)(this, JoinEagerOperation);

    var _this = (0, _possibleConstructorReturn3.default)(this, _EagerOperation.call(this, name, opt));

    _this.joinBuilder = null;
    return _this;
  }

  JoinEagerOperation.prototype.clone = function clone() {
    var _EagerOperation$proto;

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    var copy = (_EagerOperation$proto = _EagerOperation.prototype.clone).call.apply(_EagerOperation$proto, [this].concat(args));

    if (this.joinBuilder) {
      copy.joinBuilder = this.joinBuilder.clone({
        opt: copy.opt,
        expression: copy.expression
      });
    }

    return copy;
  };

  JoinEagerOperation.prototype.call = function call(builder, args) {
    var ret = _EagerOperation.prototype.call.call(this, builder, args);
    var modelClass = builder.modelClass();

    if (ret) {
      this.joinBuilder = new _RelationJoinBuilder2.default({
        modelClass: modelClass,
        expression: this.expression,
        opt: this.opt
      });
    }

    return ret;
  };

  JoinEagerOperation.prototype.onBeforeInternal = function onBeforeInternal(builder) {
    return this.joinBuilder.fetchColumnInfo(builder.knex());
  };

  JoinEagerOperation.prototype.onBeforeBuild = function onBeforeBuild(builder) {
    this.joinBuilder.build(builder);
  };

  JoinEagerOperation.prototype.onRawResult = function onRawResult(builder, rows) {
    return this.joinBuilder.rowsToTree(rows);
  };

  return JoinEagerOperation;
}(_EagerOperation3.default);

exports.default = JoinEagerOperation;
