// objection/lib/queryBuilder/QueryBuilderContextBase.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _classCallCheck2 = require("babel-runtime/helpers/classCallCheck");

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var QueryBuilderContextBase = function () {
  function QueryBuilderContextBase(userContext) {
    (0, _classCallCheck3.default)(this, QueryBuilderContextBase);

    this.userContext = userContext;
    this.options = new InternalOptions();
    this.knex = null;
  }

  QueryBuilderContextBase.prototype.clone = function clone() {
    var ctx = new this.constructor(this.userContext);

    ctx.options = this.options.clone();
    ctx.knex = this.knex;

    return ctx;
  };

  return QueryBuilderContextBase;
}();

exports.default = QueryBuilderContextBase;

var InternalOptions = function () {
  function InternalOptions() {
    (0, _classCallCheck3.default)(this, InternalOptions);

    this.skipUndefined = false;
    this.queryProps = null;
    this.debug = false;
  }

  InternalOptions.prototype.clone = function clone() {
    var copy = new InternalOptions();

    copy.skipUndefined = this.skipUndefined;
    copy.queryProps = this.queryProps;
    copy.debug = this.debug;

    return copy;
  };

  return InternalOptions;
}();
