// objection/lib/queryBuilder/operations/JoinRelationOperation.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _QueryBuilderOperation = require('./QueryBuilderOperation');

var _QueryBuilderOperation2 = _interopRequireDefault(_QueryBuilderOperation);

var _RelationJoinBuilder = require('./eager/RelationJoinBuilder');

var _RelationJoinBuilder2 = _interopRequireDefault(_RelationJoinBuilder);

var _RelationExpression = require('../RelationExpression');

var _RelationExpression2 = _interopRequireDefault(_RelationExpression);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var JoinRelationOperation = function (_QueryBuilderOperatio) {
  (0, _inherits3.default)(JoinRelationOperation, _QueryBuilderOperatio);

  function JoinRelationOperation(name, opt) {
    (0, _classCallCheck3.default)(this, JoinRelationOperation);

    var _this = (0, _possibleConstructorReturn3.default)(this, _QueryBuilderOperatio.call(this, name, opt));

    _this.expression = null;
    _this.callOpt = null;
    return _this;
  }

  JoinRelationOperation.prototype.call = function call(builder, args) {
    this.expression = _RelationExpression2.default.parse(args[0]);
    this.callOpt = args[1] || {};
    return true;
  };

  JoinRelationOperation.prototype.onBeforeBuild = function onBeforeBuild(builder) {
    var modelClass = builder.modelClass();
    var opt = (0, _assign2.default)({}, this.callOpt);

    opt.aliases = (0, _assign2.default)({}, opt.aliases);
    opt.joinOperation = this.opt.joinOperation;

    // Special case for one single relation.
    if (this.expression.numChildren === 1) {
      var relationName = void 0;

      // A bit crappy way to get the only child.
      this.expression.forEachChild(function (child, childName) {
        relationName = childName;
      });

      var relation = modelClass.getRelation(relationName);
      var alias = null;

      if (opt.alias === false) {
        alias = relation.relatedModelClass.tableName;
      } else if (opt.alias === true || !opt.alias) {
        alias = relation.name;
      } else if (_lodash2.default.isString(opt.alias)) {
        alias = opt.alias;
      }

      if (alias) {
        opt.aliases[relationName] = alias;
      }
    }

    var joinBuilder = new _RelationJoinBuilder2.default({
      modelClass: modelClass,
      expression: this.expression,
      opt: opt
    });

    joinBuilder.buildJoinOnly(builder);
  };

  return JoinRelationOperation;
}(_QueryBuilderOperation2.default);

exports.default = JoinRelationOperation;
