// objection/lib/queryBuilder/operations/RangeOperation.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _QueryBuilderOperation = require('./QueryBuilderOperation');

var _QueryBuilderOperation2 = _interopRequireDefault(_QueryBuilderOperation);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var RangeOperation = function (_QueryBuilderOperatio) {
  (0, _inherits3.default)(RangeOperation, _QueryBuilderOperatio);

  function RangeOperation(name, opt) {
    (0, _classCallCheck3.default)(this, RangeOperation);

    var _this = (0, _possibleConstructorReturn3.default)(this, _QueryBuilderOperatio.call(this, name, opt));

    _this.resultSizePromise = null;
    return _this;
  }

  RangeOperation.prototype.call = function call(builder, args) {
    if (args.length === 2) {
      var start = args[0];
      var end = args[1];

      // Need to set these here instead of `onBuild` so that they
      // don't end up in the resultSize query.
      builder.limit(end - start + 1).offset(start);
    }

    return true;
  };

  RangeOperation.prototype.onBefore = function onBefore(builder) {
    // Don't return the promise so that it is executed
    // in parallel with the actual query.
    this.resultSizePromise = builder.resultSize();
    return null;
  };

  RangeOperation.prototype.onAfter = function onAfter(builder, result) {
    return this.resultSizePromise.then(function (count) {
      return {
        results: result,
        total: parseInt(count, 10)
      };
    });
  };

  return RangeOperation;
}(_QueryBuilderOperation2.default);

exports.default = RangeOperation;
